fix(cypress): revised kill signals to processes in e2e.sh

MarceloDamian opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `master` branch of freeCodeCamp.
- [X] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any addtional description of changes below this line -->

MarceloDamian committed:
revised bash code to be more friendly

ojeytonwilliams commented on cypress/run-e2e.sh:
```diff
@@ -17,16 +17,12 @@ finally() {
   # Send kill signal to the processes
-  if [ ${#hanging_api_processes} -gt "0" ];  then
-    kill -9 $hanging_api_processes &>/dev/null
+ local process= [ ${#hanging_client_processes} -gt "0" ] && process=$hanging_client_processes || process=$hanging_api_processes  
```
Doesn't this logic mean that process is either the client or the api processes, but never both? I think it's likely that both will still be present.

Also, this currently errors out with

```
End to end bash script exiting gracefully
./cypress/run-e2e.sh: line 20: local: `[': not a valid identifier
./cypress/run-e2e.sh: line 20: local: `34': not a valid identifier
./cypress/run-e2e.sh: line 20: local: `-gt': not a valid identifier
./cypress/run-e2e.sh: line 20: local: `0': not a valid identifier
./cypress/run-e2e.sh: line 20: local: `]': not a valid identifier
```

I don't really know enough bash to say what's going on, sorry!

MarceloDamian replied:
It essentially tests if the first process isn't empty, in the case that it isn't empty it successfully stores it and continues the program skipping the "or" statement that follows. However, in the case, that it is empty the local process will be set to the next one that follows. The following if condition then checks if everything before is empty and gets set to the client process. This is under the finally() function, as long as it is called it will execute whether the promise is successful or rejected.

MarceloDamian committed:
resolved local/bash conflict

MarceloDamian commented:
I tested it locally under a test function, which I had forgotten about the first time around. I used global variables rather than local ones. Everything should be up to date. I was unable to test the "kill -9 processes", but there shouldn't be any error regarding that hopefully. Thank you so much!!!! Hope it works out!

ojeytonwilliams reviewed (changes requested):
Hi @MarceloDamian thanks for fixing the bug, but processes aren't being fully cleaned up:

MarceloDamian commented on cypress/run-e2e.sh:
```diff
@@ -11,24 +11,18 @@ finally() {
-  if [ ${#hanging_client_processes} -gt "0" ]; then
-    kill -9 $hanging_client_processes &>/dev/null
+  local process=$()
+  compare= [ ${#hanging_api_processes} -gt "0" ] && process=$hanging_api_processes || process=$hanging_client_processes  
```
Ohh!! I think I understand now! I'll convert this into a draft to work on it as soon as possible. Also thank you for review!! It means a lot.

ojeytonwilliams replied:
You're welcome.  There's no rush!

MarceloDamian committed:
revised to effectively eliminate all processes 

I included a while loop to effectively loop through all of the processes. In that way, everything gets eliminated. I tested this script locally on my machine with local variables of my creation. However, I was unable to reproduce the kill signals on my machine. Thank you so much for taking the time to review my pull request!!

MarceloDamian committed:
Update run-e2e.sh

MarceloDamian committed:
Edits for spacing

MarceloDamian committed:
Updated for indents

ojeytonwilliams commented:
Could you explain what problem this PR is intended to solve?  Unless I'm missing something this kills the same processes, but is a bit harder to understand a first glance.

MarceloDamian committed:
Updated for clarity and readability

MarceloDamian commented:
I definitely agree!! At first glance, my previous pull request wasn't really readable and a bit confusing. I recently revised an earlier draft which resembles the original code a lot! I attached the new pull request!  My PR's intention was to reduce the number of kill signals to just 1. Instead of killing the signals sequentially, it would be all killed collectively. Moving forward, in the future , this makes it easier to possibly integrate new Node processes in the scope of the while loop, rather than creating another if statement and another kill signal.
Thanks for your review!

ojeytonwilliams committed:
refactor: loop over hanging processes

ojeytonwilliams commented:
@MarceloDamian I think I get the idea now.  I've pushed a commit that simplifies things a bit, but still makes it easy to add more processes later on.  You just add them to the `processes` array and it should work.

The final `< run-e2e.sh | uniq` step was erroring out, so I cut it.

If this is okay, Marcelo, I'll ping Mrugesh and we can merge this.

MarceloDamian commented:
This is beautiful. Using arrays was ingenious and definitely more efficent. I tested it on my local machine against all cases, and they passed against all combinations (8 combinations) of hanging_api_processses, hanging_server_processses, and hanging_client_processses. I tested each case and it deleted the processes completely on my local machine. This looks perfect to me!!  Thanks!!

MarceloDamian commented on cypress/run-e2e.sh:
```diff
@@ -11,45 +11,20 @@ finally() {
   local exit_code="${1:-0}"
   # This is the clean up.
-  # Find any node processes running from within the client dir    
+  # Find any client or server node processes  
```
Would this actually be  "# Find any api, client, or server node processes" ? I was actually wondering about this before. I pushed a commit on this hopefully to clarify this. Thank you!!

ojeytonwilliams replied:
Yep, that's better, thanks for fixing it.

MarceloDamian committed:
added comment for clarification of processes

ojeytonwilliams commented:
Great! It looks like it's doing what we want. @raisedadead this is good to merge if you're happy with it.

raisedadead reviewed (approved):
Thanks for the PR @MarceloDamian and the extensive QA @ojeytonwilliams , at some point I would prefer doing this in JavaScript instead. 

I have created an issue to be able to use a node lib that can be a potential replacement to launch the e2e across packages.

MarceloDamian commented:
Yes thanks for your review guys!!
